core/data_manager.py

```diff
@@ -195,6 +195,26 @@ def add_rule(self, rule: dict, user_id: str) -> bool:
         except Error as e:
             print(e)
             return False
+
+    def edit_rule(self, rule: dict, user_id: str) -> bool:
```

Comment:
Since the SQL query here handles both adding a new rule and editing an existing rule, do we still need the `add_rule` function? How about unify them into one function called `save_rule`?

Reply:
You are right. This is much better.